docs/architecture/ADR-001-ABCI++.md

```diff
@@ -0,0 +1,414 @@
+- Malleate user transactions
+- Separate rollup block data from the stateful portion of the transaction
+- Use the appropriate signature for the selected block size
+- Follow the [non-interactive default rules](https://github.com/celestiaorg/celestia-specs/blob/master/src/rationale/message_block_layout.md#non-interactive-default-rules) (not done here)
+- Create Celestia-specific data hash by erasure-coding the block data
+
+We also need this functionality for validators to verify that
+
+- For every `MsgPayForMessage` included in the block, there is also a corresponding message and vice versa.
+- That the data hash represents the properly-erasure-coded block data for the selected block size.
+- Included messages are arranged in the expected locations in the square according to the non-interactive default rules (not done here)
+
+Technically, we don’t have to use ABCI++ yet, we could still test some needed features in the upcoming testnet without it. However, these implementations would not be representative of the implementations that would actually make it to mainnet, as they would have to be significantly different from their ABCI++ counterparts. The decision to adopt ABCI++ earlier becomes easier considering that the tendermint team has already done most of the heavy lifting, and it is possible to start working on the needed features without waiting for the cosmos-sdk team to use them. We explain our plans below to do just this, by using a subset of ABCI++ (ABCI+?) using only the new methods that are necessary, finished, and easy to incorporate into the cosmos-sdk.
+
+## Alternative Approaches
+
+While the adoption of ABCI++ is inevitable given the already made decision by upstream to implement it, here are some alternatives to the features that we need that do not use ABCI++.
+
+Alternatives for Message Inclusion
+https://github.com/celestiaorg/celestia-app/blob/92341dd68ee6e555ec6c0bb780afa3a1c8243a93/adrs/adr008:adopt-ABC%2B%2B-early.md#alternative-approaches
+
+Alternatives for Picking a square size
+https://github.com/celestiaorg/celestia-core/issues/454
+
+## Decision
+
+Proposed and initial implementation is complete.
+
+## Detailed Design
+
+### [#631](https://github.com/celestiaorg/celestia-core/pull/631) Simplified version of ABCI++ (`celestia-core`)
+
+Here we are adding only the two new methods that are necessary for the features that we need.  
+```go
+// Application is an interface that enables any finite, deterministic state machine
+// to be driven by a blockchain-based replication engine via the ABCI.
+type Application interface {
+   ...
+   PrepareProposal(RequestPrepareProposal) ResponsePrepareProposal
+   ProcessProposal(RequestProcessProposal) ResponseProcessProposal
+   ...
+}
+```
+
+It's also important to note the changes made to the request types for both methods. In upstream, they are only passing the transactions to the applications. This has been modified to pass the entire block data.
```

Comment:
One nit: can this ADR also mention that Block.Data here is slightly different from vanilla tendermint (one sentence maybe stating that we also need Evidence be part of the available data). I want the ADR to be straightforward to understand even from someone coming from vanilla tendermint and trying to see how we use abci++.

Reply:
great point https://github.com/celestiaorg/celestia-app/pull/231/commits/c9cb6063c939ac82b0a999f01f80292e6c19a2d5